### DISCUSSION TITLE
Add a way to inherit documentation <inheritdoc />

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8984] [Language Ideas]
Champion issue: https://github.com/dotnet/csharplang/issues/8985

(moved from part of https://github.com/dotnet/roslyn/issues/67 by @dfkeenan, with part of the request tracked at https://github.com/dotnet/csharplang/issues/315)

Hi,

**Please add `<inheritdoc />` tag or similar:**

I think it would be really handy if the compiler(s) when outputting the documentation XML file if it could use inherited documentation where available/requested.  Though I am not sure what should be output if there is no documentation available.

P.S. A Diagnostic and Code Fix that generates document comments for me would also great :smiley:

**Comment by daveaglick** [comment]
Should `<inheritdoc />` and namespace comments be tracked in separate issues?

Also, FWIW, I have an open pull request at https://github.com/dotnet/roslyn/pull/15494 that addresses namespace documentation comments as described in https://github.com/dotnet/roslyn/issues/15474 (which hasn't been closed or migrated yet).

**Comment by gafter** [comment]
@daveaglick Yes, having separate issues would be great. If you create a new one for one of these and link to it here, I'll edit the OP here.

**Comment by daveaglick** [comment]
@gafter :+1:

I created https://github.com/dotnet/csharplang/issues/315 which is just a copy of https://github.com/dotnet/roslyn/issues/15474 (so I guess that corresponding Roslyn issue can also be closed now)

**Comment by terrajobst** [comment]
Is's worth mentioning that this tag can be (and already is) supported by documentation generators. By making it a Roslyn feature it makes it first class by making IntelliSense work (e.g. hovering over calls in the class overriding the member shows the documentation). Also, it should be implied to reduce boilerplate without causing a compiler warning for missing documentation.

**Comment by sharwell** [comment]
I believe a good way to start here would be examining the features currently provided by SHFB for this documentation element, and then identifying ones which would be *impractical* from an implementation perspective or "not particularly needed" from an overall usage perspective.

:link: [inheritdoc](http://ewsoftware.github.io/XMLCommentsGuide/html/86453FFB-B978-4A2A-9EB5-70E118CA8073.htm) (Sandcastle XML Comments Guide)

## Syntax

I believe it's reasonable to use the same form as defined by SHFB:

```xml
<inheritdoc [cref="member"] [select="xpath-filter-expr"] />
```

This element may be placed either as a top-level element or as an inline element in a documentation comment.

## Compilation

The compiler is updated in the following ways to account for this element:

* The compiler MUST allow the use of an `inheritdoc` element as a top-level element in XML documentation comments
* The compiler MUST allow the use of an `inheritdoc` element as an inline element in XML documentation comments
* The compiler SHOULD report a warning if the `inheritdoc` element appears without a `cref` attribute, and no candidate for inheriting documentation exists
* The compiler MUST evaluate the `cref` attribute value in the same manner as it does for the `see` element, and include the resolved documentation ID as the value of the attribute in the output documentation file
* The compiler MAY report a warning if the `select` attribute is specified, but the value is not a syntactically valid XPath expression
* The compiler SHOULD preserve the placement and form of the `inheritdoc` element in the XML documentation file. It MUST NOT replace the `inheritdoc` element with inherited documentation when writing the documentation file. (Edit: This one is the subject of disagreement below; I plan to revisit after writing up a proposal for the Tools section.)

### Candidate for inheritance

⚠️ Some items in this list are written as existence (Boolean), while others identify the actual candidate(s). This section should be cleaned up with the understanding that the compiler only needs to care about existence for the purpose of reporting a warning as described above.

* For types and interfaces
  * The type is derived from a class which is not `System.ValueType`, `System.Enum`, `System.Delegate`, or `System.MulticastDelegate`, OR
  * The type implements a named interface which is not itself
* For constructors
  * The containing type of the constructor is derived (directly or indirectly) from a type which contains a constructor with the same signature
* For explicit interface implementations
  * The candidate is the implemented interface member
* For overriding methods (methods marked with `override`
  * The candidate is the overridden method
* For implicit interface implementations
  * The candidates are the implemented interface members

### Impacted warnings

The following warnings, which are specific to the Roslyn implementation of a compiler for C#, should be updated to account for this feature:

#### CS1573

> Parameter 'parameter' has no matching param tag in the XML comment for 'parameter' (but other parameters do)

Care should be taken to not report this warning in the following scenario. It would be sufficient to disable this warning any time `inheritdoc` appears as a top-level documentation element.

```csharp
class Base {
  /// <param name="x">Doc for x</param>
  protected void Method1(int x) { }
}

class Derived : Base {
  /// <inheritdoc cref="Base.Method1"/>
  /// <param name="y">Doc for y</param>
  protected void Method2(int x, int y) { }
}
```

#### CS1712

> Type parameter has no matching typeparam tag in the XML comment (but other type parameters do)

This is similar to the previous case, but applies for type parameters. The example shows type parameters for a generic type, but it can also apply to generic methods.

```csharp
/// <typeparam name="T">Doc for T</param>
class Base<T> {
}

/// <inheritdoc/>
/// <typeparam name="T2">Doc for T2</param>
class Derived<T, T2> : Base<T> {
}
```

## Tools

This section describes the manner (semantics) in which `inheritdoc` elements should be evaluated.

*Will be added in a later comment*

**Comment by pascalberger** [comment]
> The compiler SHOULD preserve the placement and form of the inheritdoc element in the XML documentation file. It MUST NOT replace the inheritdoc element with inherited documentation when writing the documentation file.

Why shouldn't `inheritdoc` be replaced in the XML documentation file? This would make the generated XML documentation files usable to a much broader set of tools IMHO

**Comment by daveaglick** [comment]
I agree with @pascalberger above about replacing in the XML documentation file. There are tools, including documentation generators, that use the XML doc file as a substitute for having access to the source. If the `inheritdoc` comments aren't replaced in the output XML file then those tools will have to continue to implement the `inheritdoc` logic and we're back to where we started.

If that's not feasible, perhaps the `XmlDocumentationProvider` could be extended to perform the substitution. At least that way a tool/generator could read the assembly with Roslyn, hooking up the XML comment file via the provider, and get consistent `inheritdoc` resolution.

**Comment by sharwell** [comment]
> Why shouldn't `inheritdoc` be replaced in the XML documentation file?

Ignoring the "it's what we do now" issue, three immediate reasons:

1. The XML documentation files might not be available at compilation time, but added later. For example, suppose X.dll 1.0 depends on Y.dll 1.0, but the latter doesn't include XML documentation. A user later builds App.exe depending on X.dll 1.0 and explicitly updates Y.dll to 1.1, which now includes documentation. The inherited documentation should propagate to methods in X.dll where `inheritdoc` was specified.
1. XML documentation of a base type may change, and typically users want the changes to propagate  without requiring the library be recompiled. (I know in all cases where I used `inheritdoc` in the past, this was my expectation.)
1. Duplicating the documentation unnecessarily bloats the output XML form.

> At least that way a tool/generator could read the assembly with Roslyn, hooking up the XML comment file via the provider, and get consistent `inheritdoc` resolution.

I would expect Roslyn to provide an API which provides code element documentation with `inheritdoc` expanded.

**Comment by pascalberger** [comment]
While I see the advantages you mention, another downside, beside documentation generators mentioned by @daveaglick, of this approach is that if a library is created with `inheritdoc` tags in the documentation XML they won't be resolved for users of older versions of Visual Studio.

**Comment by sharwell** [comment]
> While I see the advantages you mention, another downside, beside documentation generators mentioned by @daveaglick, of this approach is that if a library is created with `inheritdoc` tags in the documentation XML they won't be resolved for users of older versions of Visual Studio.

That's a very good point. I forgot I use SHFB as a preprocessor to resolve these before packaging NuGet from projects that use inheritdoc.

:link: https://github.com/tunnelvisionlabs/dotnet-threading/commit/a253a2b2e5ba0ef9614a978c89a32ad2d6d08804

**Comment by matkoch** [comment]
Given the potential pitfalls, I'm very much against adopting the `select` feature. When changing the documentation in some base class, there is no easy way to verify that none of the derived types won't break (i.e., gets their documentation changed unintentionally).

**Comment by sharwell** [comment]
> Given the potential pitfalls, I'm very much against adopting the `select` feature. When changing the documentation in some base class, there is no easy way to verify that none of the derived types won't break (i.e., gets their documentation changed unintentionally).

Do you have examples? I've only *very* rarely needed `select`, but never run into specific problems with it. However, it's presence makes it a bit easier (IMO) to explain the behavior of `inheritdoc` as a nested element.

**Comment by matkoch** [comment]
@sharwell you could, for instance, select the _second paragraph_ via `select`, and if you later insert a paragraph between the first and the second, the inherited paragraph gets changed without noticing that.

**Comment by daveaglick** [comment]
I'll also ++ for no `select`. In addition to the problems @matkoch mentioned with relative positioning of content, I think it also makes the feature more complex and confusing for very little actual gain.

**Comment by sharwell** [comment]
@matkoch I've never seen someone do that. Considering that it's hard to get users to write *any* documentation, much less put thought into the specific content resulting from what they write, I do not anticipate any notable number of problems arising from this feature. The only thing I've ever used it for IIRC was to remove a "note to implementers" that would otherwise have been inherited. It was a strange message to otherwise be included in the documentation for a sealed method, and I was happy to have a way to remove it.

**Comment by matkoch** [comment]
@sharwell it is even used as an example in the [Sandcastle documentation](https://ewsoftware.github.io/XMLCommentsGuide/html/86453FFB-B978-4A2A-9EB5-70E118CA8073.htm) (`select="para[last()]"`). In my opinion it shouldn't be a reasoning about whether someone is actually using this or not. It's about locality, the potential defects a change can introduce and the (understandable) missing tooling. If you add a new abstract method or change some signature, you'll get a compiler error, since derived types are affected. Changing inherited documentation however would become a task of manual verification.

I also agree with @daveaglick. Too much complexity for little gain.

**Comment by matkoch** [comment]
@sharwell A compromise could be to allow only a restricted set of select expressions. Like selection by ID.

**Comment by sharwell** [comment]
Given I have had use for `select`, I lean in favor of having the option. However, given how rarely I've needed it, it's certainly not essential.

> A compromise could be to allow only a restricted set of select expressions. Like selection by ID.

This would be bad. Given we're already in the context of XML, XPath has a specific meaning and we don't have to redefine it. If we allow the attribute but don't use XPath, we have to redefine everything and can't use an off-the-shelf XPath implementation.

**Comment by matkoch** [comment]
@sharwell I didn't mean to introduce a new syntax, but rather to allow only a subset of xpath. I.e., no `last()` or selection by index.

**Comment by sharwell** [comment]
@pascalberger How about this? (I only modified the main affected element; other changes way be needed elsewhere in the text for consistency if this approach is adopted.)

> The compiler SHOULD emit the documentation file with `inheritdoc` elements replaced by their inherited content. The compiler MAY support emitting the documentation file with `inheritdoc` elements not replaced; in this case the compiler SHOULD preserve the placement and form of the `inheritdoc` element in the XML documentation file, except with the `cref` attribute expanded to the documentation ID of the referenced member.

**Comment by sharwell** [comment]
## Tools

:memo: This section defines the inheritance semantics of the `inheritdoc` element, as they would apply to the tools responsible for interpreting the meaning of `inheritdoc`. If the compiler expands `inheritdoc` during the production of a documentation file (e.g. https://github.com/dotnet/csharplang/issues/313#issuecomment-322247608), then the compiler would be considered a tool for the purposes of this section.

### General

The expansion of an `inheritdoc` element produces a XML node set which replaces the `inheritdoc` element.

### Inheritance rules

The inheritance rules determine the element(s) from which documentation is inherited. The behavior is unspecified if a cycle exists in these elements.

The search order is defined under the **Top-Level Inheritance Rules**
 section of [inheritdoc](http://ewsoftware.github.io/XMLCommentsGuide/html/86453FFB-B978-4A2A-9EB5-70E118CA8073.htm). However, the rules for determining which elements to ignore are generalized to the following:

* The `overloads` element is never inherited when the `select` attribute is omitted. In other words, the default value for the `select` attribute is `*[not(self::overloads)]`.
* If an element includes a `cref` attribute, it is only omitted if the matching existing element has a `cref` attribute that resolves to the same documentation ID is already included
* If an element includes a `href` attribute, it is only omitted if the matching existing element has an `href` attribute with an equivalent URI
* If an element includes a `name`, `vref`, and/or `xref` attribute, it is only omitted if the matching existing element has the corresponding attribute(s) with the same value(s)
* After observing the above, an element is omitted if an existing element (already inherited or a sibling of the `inheritdoc` element) has the same element name

### Inline `inheritdoc` elements

When an `inheritdoc` element appears inline (as opposed to the top level), the base node from which the `select` query is evaluated changes to the parent element of the `inheritdoc` element. The path to the matching node is identified by all of the following:

* The element name
* The values of attributes of the element
* The index of the element

For example, the following `inheritdoc` elements are equivalent:

```csharp
/// <summary>
/// <list type="number"><item></item></list>
/// <list type="bullet">
/// <item></item>
/// <item><inheritdoc/></item>
/// </list>
class WithoutSelectAttribute { }

/// <summary>
/// <list type="number"><item></item></list>
/// <list type="bullet">
/// <item></item>
/// <item><inheritdoc select="/summary[0]/list[@type='bullet'][0]/item[1]/*[not(self::overloads)]"/></item>
/// </list>
class WithSelectAttribute { }
```

**Comment by pascalberger** [comment]
> The compiler SHOULD emit the documentation file with inheritdoc elements replaced by their inherited content. The compiler MAY support emitting the documentation file with inheritdoc elements not replaced; in this case the compiler SHOULD preserve the placement and form of the inheritdoc element in the XML documentation file, except with the cref attribute expanded to the documentation ID of the referenced member.

@sharwell Does this mean the compiler is free how it implements it (either the SHOULD or the MAY sentence)? Or that it's some kind of an option?

Otherwise LGTM

**Comment by sharwell** [comment]
> Does this mean the compiler is free how it implements it (either the SHOULD or the MAY sentence)? Or that it's some kind of an option?

@pascalberger It's written with the intent of acknowledging users may want an option to control this, while also suggesting the preferred behavior in the event the compiler only implements one or the other.

**Comment by matkoch** [comment]
@sharwell I'm confused now given your recent post. Does that mean that you will keep the `select` feature?

**Comment by sharwell** [comment]
@matkoch I'm in favor of keeping select, but wanted to present the alternative in concrete wording for the sake of continued discussion. As mentioned in the alternative wording comment, other language including that seen in the Tools section would need to be updated to reflect the final supported behavior.

**Comment by matkoch** [comment]
Anyone got an idea how this following example should be handled?

    public interface IFoo
    {
      event EventHandler A;
      event EventHandler B;
    }
    public class Bar : IFoo
    {
      /// <inheritdoc />
      public event EventHandler A, B;
    }

Will it require a `cref` definition? And if yes, will we always allow to reference symbols, not contained in the actual _inheritance chain_ ?

**Comment by sharwell** [comment]
@matkoch It would be treated exactly the same as if the comment were separately applied to `Bar.A` and `Bar.B`.

**Comment by dotMorten** [comment]
There's some chat between having the xml doc used for intellisense vs for generating documentation, supporting older clients etc. Perhaps the real solution here is to generate two? One for intellisense with the inheritdoc content evaluated, internals, privates, remarks etc stripped out*, , and one for generating doc with everything that you don't actually ship with the binary?

(*Stripping out privates, internals, remarks etc not used for intellisense can significantly reduce the payload for the intellisense doc)

**Comment by sharwell** [comment]
I've started work on an initial implementation and have some additional notes:

1. For consistency with `<include>`, I believe the `select` attribute in the proposal above should be renamed to `path`. @EWSoftware this is likely to cause a discrepancy with Sandcastle, but I believe this aspect of `inheritdoc` is "relatively" infrequently used and I'm hoping it will not be overly burdensome to update SHFB to support `path` (with `select` retained as a compatibility fallback).
    * The new behavior could result in an unexpected change in output for users who expected `select` to work with the C# compiler. We should consider either supporting `select` as an alias for `path`, reporting a warning for `select`, or aiding in the implementation of an analyzer/code fix that reports diagnostics for `select` and offers to convert it to `path` instead.
1. Given the compiler already uses XPath for `<include>` elements, it seems obvious to support the same for `<inheritdoc>`. Given that the compiler will *validate* the included XML content, it may be necessary to use filters in some cases to exclude content which would result in compilation warnings. (Whether that takes the form of a `path` attribute or a `select` attribute is a separate question.)
1. It's unclear (to me) if the new behavior should be tied to `LangVersion` or exposed as a compiler option in some manner. In the current prototype, it's simply implemented in the same manner as `include` elements are handled, meaning the replacement is always enabled during builds.

**Comment by EWSoftware** [comment]
@sharwell It's a minor change in two places in the SHFB tool to also check for `path` if `select` isn't found.  I could also have the tool issue a warning if `select` is found stating that it's deprecated and to use `path` instead.  Once this is ironed out and ready for implementation, let me know and I can make the appropriate changes.

**Comment by krdmllr** [comment]
Is this blocked because we wait for https://github.com/dotnet/csharplang/pull/1749#pullrequestreview-141630318 ?

**Comment by pi3k14** [comment]
Could you please give an update to the thousand of users requesting this? 
ref. https://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/3745102-add-intellisense-support-for-the-inheritdoc-ta

**Comment by gafter** [comment]
> While I see the advantages you mention, another downside, beside documentation generators mentioned by @daveaglick, of this approach is that if a library is created with `inheritdoc` tags in the documentation XML they won't be resolved for users of older versions of Visual Studio.

Users of older versions of Visual Studio are probably using older runtime platforms, which don't use `inheritdoc` in their APIs.  They are also using an older compiler, which doesn't understand it.

**Comment by marksmeltzer** [comment]
@gafter,

Are you saying that VS 2019 will show the inherited docs in the intellisense? 2017 still doesn't (unless there's an update I've missed).

Regards,

**Comment by pascalberger** [comment]
> Users of older versions of Visual Studio are probably using older runtime platforms, which don't use inheritdoc in their APIs. They are also using an older compiler, which doesn't understand it.

@gafter My point you quoted was that if we have the possibility to replace `inheritdoc` we can use `inheritdoc` in a library, using a recent compiler which supports it, while the consuming project can still use an older compiler, as all `inheritdoc` are resolved (analog to SHFB preprocessing functionallity)

**Comment by dotMorten** [comment]
> Users of older versions of Visual Studio are probably using older runtime platforms, which don't use inheritdoc in their APIs. They are also using an older compiler, which doesn't understand it.

That's a good approach to never move forward. Or we could just move forward.

Alternatively, generate the XML doc with the inherit doc pieces resolved during compilation.

**Comment by gafter** [comment]
I believe the best approach is to have APIs for consuming documentation XML that are capable of handling the inherited documentation.  Older tools may have to migrate to using the new APIs before clients get the benefits, but I think this is the right solution.

**Comment by poke** [comment]
During .NET Conf, in [this video](https://youtu.be/mg2aocZA9Pc?t=1540), Kendra and Mika showed off some `<inheritdoc />` goodness that is apparently coming with the next VS preview. So is this suddenly coming after all? That’s awesome! 😄

**Comment by Joe4evr** [comment]
> During .NET Conf, in [this video](https://youtu.be/mg2aocZA9Pc?t=1540), Kendra and Mika showed off some `<inheritdoc />` goodness that is apparently coming with the next VS preview. So is this suddenly coming after all? That’s awesome! 😄

Yep, according to [the release notes of the latest VS update](https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes-preview#16.4.0-pre.1.0):
> Methods that have no XML documentation can now automatically inherit XML documentation from the method it is overriding. Place your cursor over the undocumented method that implements a documented interface method. Quick Info will then display the XML documentation from the interface method.

**Comment by Alxandr** [comment]
So this is implicit, and not explicit? Meaning, that if you do use `<inheritdoc />`, it's not shown? And what about if you build it? Does all the documentation generators now have to update so that derived methods not documented should include their parent docs?

**Comment by Joe4evr** [comment]
> So this is implicit, and not explicit? Meaning, that if you do use `<inheritdoc />`, it's not shown? And what about if you build it? Does all the documentation generators now have to update so that derived methods not documented should include their parent docs?

That's definitely a small concern. At this particular time, it *seems* that this is purely implemented as part of VS and not the compiler? If that's the case, that means the compiler-generated doc file would still have a blank entry for undocumented members. Also, lack of `<inheritdoc />` would mean that tools like DocFX wouldn't copy over the base-descriptions, unless they update to this behavior which is possibly not entirely desirable for everyone.

**Comment by daveaglick** [comment]
> unless they update to this behavior which is possibly not entirely desirable for everyone

No, but I could see making it an option - especially now that VS users will gradually become conditioned to seeing implicit doc inheritance in the IDE. A generator could present implicitly inherited documentation with some sort of graphical indicator or in a different style to indicate the difference 🤔.

**Comment by sharwell** [comment]
> So this is implicit, and not explicit?

Both were added to the IDE. This allows the feature to benefit the most common documentation scenarios that previously did not show any documentation to users.

> Does all the documentation generators now have to update so that derived methods not documented should include their parent docs?

At this time, yes. There are no plans I'm aware of to make the compiler itself understand `<inheritdoc/>`. Roslyn inside Visual Studio is one tool that consumes documentation where members may be missing documentation and/or have an `<inheritdoc/>` element in the documentation, and it chooses to interpret this documentation in a particular way. Other tools would be responsible for their own implementations of the same features, if desired.

Keep in mind the overwhelming majority of projects are not using documentation post-processing tools, and the few that are will have a few different paths forward to realize the new feature. One possible path forward is to use an analyzer and code fix that automatically adds `/// <inheritdoc/>` to elements which are currently missing documentation.

**Comment by dotMorten** [comment]
When you use `<inheritdoc/>` it also goes into the intellisense xml file. Will intellisense now also work for libraries that uses that in their xml doc?

**Comment by sharwell** [comment]
> Will intellisense now also work for libraries that uses that in their xml doc?

It should, yes 👍

**Comment by linvi** [comment]
@sharwell 

> Both were added to the IDE. This allows the feature to benefit the most common documentation scenarios that previously did not show any documentation to users.

When you say IDE I assume you mean VS 2019 only? Or can we hope for some C# plugin of VS Code / VS 2017 also?

Is there an ETA for the new version of IDE that will include this feature?

**Comment by sharwell** [comment]
@linvi It's a Roslyn feature, so each of the IDEs that use Roslyn for IntelliSense should have this feature. Roslyn 3.4 contains the feature, and will first ship with Visual Studio 2019 version 16.4 (I don't have a specific date). For other IDEs the specific release version/date would be up to that specific project.

Documentation post-processing tools would need to be updated separately. Sandcastle Help File Builder already supports the feature, but other tools may need changes.

**Comment by linvi** [comment]
@sharwell Thanks for the answer,

Do you know if the IDEs will be by default consider that the tag `/// <inheritdoc />` is set on all methods/class/interfaces.
I am asking as this is already the case with `Jetbrains Rider IDE`.

**Comment by EWSoftware** [comment]
@sharwell Back in the [Jul 29th post](https://github.com/dotnet/csharplang/issues/313#issuecomment-408695215) you were contemplating using `path` rather than `select`.  Which did you choose?  If it was `path`, I'll make the noted changes in SHFB to support it.

**Comment by sharwell** [comment]
@linvi _Members_ treat it that way by default, but _types_ do not.

@EWSoftware I used `path` 👍

**Comment by saucecontrol** [comment]
Since there are no current plans to have the compiler replace `<inheritdoc />` elements in the output XML, I put together an MSBuild task that will do it just after compile.  It's here if anyone would like such a thing: https://github.com/saucecontrol/InheritDoc

**Comment by gulshan** [comment]
Is this done?
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/inheritdoc

**Comment by saucecontrol** [comment]
> Is this done?
> https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/inheritdoc

Yep, starting from Visual Studio 16.4

**Comment by albert-github** [comment]
Is there a reason why `<inheritdoc.../>` is not in the https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md yet?

**Comment by 333fred** [comment]
`<inheritdoc>` is a compiler feature, not a language feature. It won't go in the specification because it doesn't belong there.

**Comment by 333fred** [comment]
@BillWagner or @sharwell, do we have this documented somewhere?

**Comment by EWSoftware** [comment]
Unofficially it's documented in the [Sandcastle XML Comments guide](http://ewsoftware.github.io/XMLCommentsGuide/html/86453FFB-B978-4A2A-9EB5-70E118CA8073.htm).  It's also documented in the Microsoft Docs [C# programming guide](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/inheritdoc) but isn't as thorough and doesn't mention the `path` attribute.

**Comment by albert-github** [comment]
@333fred You mentioned: it is a compiler feature. But it is in the comment and describing the some documentation, wouldn't this typically belong to the mentioned chapter in the manual? This chapter is called "Documentation comments" and it states at the first line: "C# provides a mechanism for programmers to document their code using a special comment syntax that contains XML text. "

**Comment by 333fred** [comment]
> But it is in the comment and describing the some documentation, wouldn't this typically belong to the mentioned chapter in the manual?

The manual you linked is the C# language specification, defined as ECMA-334. `<inheritdoc>` is _not_ a C# language feature, it is a special tag that Roslyn/docfx understands. It is therefore not part of, nor will it be part of, ECMA-334, unless someone champions the feature to add it to the formal definition of the language. It may seem like me being pedantic, but it's the difference between "a 3rd party C# compiler needs to implement this feature in order to be considered a C# compiler" and "a 3rd party C# compiler might want to implement this so that existing code that works with Roslyn and docfx works with it."

**Comment by albert-github** [comment]
I linked to  https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md  in my first comment (https://github.com/dotnet/csharplang/issues/313#issuecomment-763507911) and in my second comment (https://github.com/dotnet/csharplang/issues/313#issuecomment-763908513) I made reference of the quoted which is in the mentioned link as well as in the Annex D of the:

![ecma](https://user-images.githubusercontent.com/5223533/105335141-f9160c00-5bd7-11eb-854a-6a17573bed7a.png)

so I would expect that the `<inheritdoc>` will be in the Annex D of the forthcoming standard and thus in the linked file.

I'm definitely not saying it is part of the c# language but I don't think `<see>` is part of the language though it is described in this Annex D.

**Comment by 333fred** [comment]
> I'm definitely not saying it is part of the c# language but I don't think `<see>` is part of the language though it is described in this Annex D.

If it is part of the spec, it is part of the language.

**Comment by albert-github** [comment]
I don't really see a difference between `<see>` and `<inheritdoc>`, must be my lack of knowledge about what is part of the C# language and what not.

**Comment by sharwell** [comment]
@albert-github The `<see>` element was reviewed and approved for inclusion in the language a long time ago. The `<inheritdoc>` element is not a `<see>` element, so it would need to be reviewed separately if it was to be included. Currently, neither the C# compiler nor the language specification has any knowledge of the `<inheritdoc>` element. If one appears in source code, it is treated as an unknown documentation element that is passed on to documentation processing tools unchanged, where they are free to either ignore it or use it.

**Comment by albert-github** [comment]
@sharwell This shines a bit of new light on the problem.

- The following is my formulation: so you are saying that `<see>` entered in the language in the past but that this nowadays wouldn't directly happen as `<see>` is also just for documentation.
- where will there be an authoritative source for documentation tags (as the `<inheritdoc>` is now unofficially documented by Sandcastle and Microsoft as mentioned in https://github.com/dotnet/csharplang/issues/313#issuecomment-763869846)
- having the discrepancy between the ECMA specification and the unofficial description at Sandcastle and Microsoft doesn't make it very clear to a normal user, even though in the Annex D there are some remarks about ignoring tags, like:
  > A conforming C# compiler is not required to check the syntax of documentation comments; such
comments are simply ordinary comments. A conforming compiler is permitted to do such checking,
however.
  >
  > The documentation generator must accept and process any tag that is valid according to the rules of XML.
The following tags provide commonly used functionality in user documentation. (Of course, other tags are
possible.)
- what is necessary to have `<inheritdoc>` in the ECMA standard? Most likely Sandcastle and / or Microsoft have to start a review path for this tag (and maybe there are others unknown to me).

**Comment by poke** [comment]
The correct place to expand the documentation would probably be the [XML documentation chapter on this](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/inheritdoc), i.e. the following document on GitHub: https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/xmldoc/inheritdoc.md.

**Comment by albert-github** [comment]
@poke Looks like promising, though I don't see any reference that this is a repository intended for preparing for standardization by e.g. ECMA  / ISO / ...

**Comment by 333fred** [comment]
> though I don't see any reference that this is a repository intended for preparing for standardization by e.g. ECMA / ISO / ...

This repository is about additions to the C# language. Eventually, such changes will make their way into the ecma spec, as they are part of the language.

**Comment by albert-github** [comment]
@333fred Nearly contradicting to previous comments, but as I understand it now the `<inhertitdoc>` might in the future be part of the ECMA standard as part of the language, hopefully it won't take too long.

**Comment by artdedeco** [comment]
I'm not sure if this is the right place; Sorry if it's not.

I've noticed that `<inheritdoc />` now has greater priority over docs defined in the member itself. I remember this not being case a couple years back. 

Eg.:

```cs
        /// <summary>a</summary>
        public void A() {}

        /// <inheritdoc cref="A()"/>
        /// <summary>b</summary>
        /// <returns>foo</returns>
        public void B() {}
```

The output docs for `B` is:

```
a

Returns:
foo
```

When I expected:

```
b

Returns:
foo
```

**Comment by CyrusNajmabadi** [comment]
tagging @sharwell .  however, it's not a case where one set of docs is brought in, then overwritten by other docs.  Rather, the set of docs brought in is just brought in verbatim with no processing done on it.

**Comment by ZacharyPatten** [comment]
https://github.com/dotnet/roslyn/issues/54494 @andre-ss6

**Comment by geoffbon** [comment]
Is it intended that `<inheritdoc />` be usable to embed inside parameter descriptions?

For example, a constructor might take values to populate the class properties, and the parameters of that constructor will have the same description as the original property.

For example:
```cs
class MyEntity
{
       /// <summary>
        /// The code of the item
        /// </summary>
        public string Code
        {
            get;
            set;
        }

       /// <summary>
        /// The description of the item
        /// </summary>
        public string Description
        {
            get;
            set;
        }

        /// <summary>
        /// Creates a new entity
        /// </summary>
        /// <param name="code"><inheritdoc cref="MyEntity.Code"/>/param>
        /// <param name="description"><inheritdoc cref="MyEntity.Description"/>/param>
        public MyEntity(string code, string description)
        {
        }
}
```
Currently this example inserts blank values in place of the `<inheritdoc />`  tag.

**Comment by ZacharyPatten** [comment]
@geoffbon 

It was intended that `inheritdoc` copies the XML docs from the source at the same XML path as it is used on the target. However, there is a bug in it currently where it concatenates docs rather than copying the specific doc at the identical x-path.

In your case you are using `inheritdoc` inside the `param` so it would copy the content from the source that is at the same x-path (also inside `param` tag). However you also have malformed XML. You are missing a `<` on `/param>`.

Personally, I don't like how it functions based on x-paths by default, but it was intended.

**Comment by ZacharyPatten** [comment]
sorry I bumped the comment button earlier than intended... Here is an example that will fix your issue @geoffbon 

```cs
class MyEntity
{
	/// <summary>
	/// The code of the item
	/// </summary>
	public string Code
	{
		get;
		set;
	}

	/// <summary>
	/// The description of the item
	/// </summary>
	public string Description
	{
		get;
		set;
	}

	/// <summary>
	/// Creates a new entity
	/// </summary>
	/// <param name="code"><inheritdoc cref="Code" path="/summary"/></param>
	/// <param name="description"><inheritdoc cref="Description" path="/summary"/></param>
	public MyEntity(string code, string description)
	{
	}
}
```
If you explicitly provide an x-path (via the `path` attribute on the `inheritdoc` tag) you can override the path used during the look up.

**Comment by geoffbon** [comment]
This is outstanding, thanks so much @ZacharyPatten!  This makes a lot more sense to me now.

**Comment by Varorbc** [comment]
vs 17.4 has added the inherit doc feature, but what is the shortcut?

**Comment by binki** [comment]
@Varorbc what sort of shortcut are you looking for? If you want IntelliSense to inherit the full documentation for an interface member or abstract method, it already does that if you simply omit the documentation comment entirely.
